DOC: Add details of dropna in DataFrame.pivot_table

it176131 opened the pull request:
- [x] closes #61113 
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

it176131 committed:
modified:   pandas/tests/reshape/test_pivot.py
	- Added test :callable:`test_pivot_table_index_and_column_keys_with_nan` to test that null index/column keys are kept when `dropna=False`, and removed when `dropna=True`.

it176131 committed:
modified:   pandas/core/reshape/pivot.py
	- Updated docs for `dropna` parameter in :callable:`pivot_table`.

it176131 commented:
Which `doc/source/whatsnew/vX.X.X.rst` file should I edit? `v2.3.0.rst`?

it176131 committed:
modified:   pandas/tests/reshape/test_pivot.py
	- Replaced `np.NaN` with `np.nan` to pass failing tests on CI build.

it176131 committed:
modified:   pandas/core/reshape/pivot.py
	- Added comma and period to `dropna` parameter in :callable:`pivot_table` to pass docstring validation hook in CI build.

rhshadrach commented on pandas/core/reshape/pivot.py:
```diff
@@ -102,8 +102,9 @@ def pivot_table(
         Do not include columns whose entries are all NaN. If True,
-        rows with a NaN value in any column will be omitted before
-        computing margins.
+        - rows with a NaN value in any column will be omitted before computing margins,
+        - index/column keys containing NA values will be dropped (see ``dropna``
+        parameter in ``pandas.DataFrame.groupby``).
```
You'll also need to update the entry in `shared_docs` within `pandas.core.frame`.

rhshadrach replied:
Can you use `an NA value` instead of `a NaN value`. Also link to the groupby via `` :meth:`DataFrame.groupby` `` instead of ``pandas.DataFrame.groupby``.

rhshadrach replied:
It's `:meth:` and not `:method:` for sphinx.

rhshadrach commented on pandas/tests/reshape/test_pivot.py:
```diff
@@ -2529,6 +2529,63 @@ def test_pivot_table_aggfunc_nunique_with_different_values(self):
 
         tm.assert_frame_equal(result, expected)
 
+    def test_pivot_table_index_and_column_keys_with_nan(self, dropna: bool) -> None:
```
Have you checked that this is not already tested?

it176131 replied:
I looked at the other `pivot_table` tests that used `dropna` but didn't see any explicitly calling out the index/column keys as NA with all other parameters set to default. [This one](https://github.com/pandas-dev/pandas/blob/b64f438cc8079d441331396fbac1e2dc61b26af9/pandas/tests/reshape/test_pivot.py#L235) is the closest, but it sets `observed=False` instead of `observed=True` (default).

rhshadrach reviewed (changes requested):
Thanks for the PR!

> Which `doc/source/whatsnew/vX.X.X.rst` file should I edit? `v2.3.0.rst`?

No need to add to the whatsnew when improving a docstring.

it176131 committed:
modified:   pandas/core/reshape/pivot.py
	- Updated reference link to `pandas.DataFrame.groupby` with :method: syntax in the :callable:`pivot_table` param `dropna` docstring.

it176131 committed:
modified:   pandas/core/reshape/pivot.py
	- Replaced "NaN" with "NA" per request from @rhshadrach in comment.

it176131 committed:
modified:   pandas/core/frame.py
	- Updated value of :attr:`DataFrame._shared_docs["pivot_table"]` to include updated docstring from `pandas.core.reshape.pivot.pivot_table`.

rhshadrach commented on pandas/tests/reshape/test_pivot.py:
```diff
@@ -2529,6 +2529,63 @@ def test_pivot_table_aggfunc_nunique_with_different_values(self):
+    def test_pivot_table_index_and_column_keys_with_nan(self, dropna: bool) -> None:
+        """Index/column keys containing NA values depend on ``dropna``.
+
+        Input data
+        ----------
+            row  col  val
+        0  NaN  0.0    0
+        1  0.0  1.0    1
+        2  1.0  2.0    2
+        3  2.0  3.0    3
+        4  3.0  NaN    4
```
Can you remove this docstring. Employing this pattern throughout tests would be a large maintenance burden and in my opinion is not be sustainable

rhshadrach commented on pandas/tests/reshape/test_pivot.py:
```diff
@@ -2529,6 +2529,63 @@ def test_pivot_table_aggfunc_nunique_with_different_values(self):
+        """
+        data = {"row": [None, *range(4)], "col": [*range(4), None], "val": range(5)}
+        df = DataFrame(data)
+        actual = df.pivot_table(values="val", index="row", columns="col", dropna=dropna)
```
Can you call this `result` instead of `actual`.

rhshadrach commented on pandas/tests/reshape/test_pivot.py:
```diff
@@ -2529,6 +2529,63 @@ def test_pivot_table_aggfunc_nunique_with_different_values(self):
+            data=e_data,
+            index=Index(data=e_index, name="row"),
+            columns=Index(data=e_columns, name="col"),
+        ).sort_index()
```
Can you setup the data exactly as expected instead of sorting; in general applying manipulations to the expected result makes tests harder to understand. Recommend using `range(2)` instead of `range(4)` above; we get the same test coverage, but much easier to manage.

rhshadrach commented on pandas/tests/reshape/test_pivot.py:
```diff
@@ -2529,6 +2529,63 @@ def test_pivot_table_aggfunc_nunique_with_different_values(self):
+        ).sort_index()
+        if dropna:
+            expected = (
+                # Drop null index/column keys.
+                expected.loc[expected.index.notna(), expected.columns.notna()]
+                # Drop null rows.
+                .dropna(axis="index", how="all")
+                # Drop null columns.
+                .dropna(axis="columns", how="all")
```
Similar to the above; I think it'd be okay here to use `.loc` with explicit indices, but we should not be using `notna` and `dropna`.

rhshadrach commented on pandas/tests/reshape/test_pivot.py:
```diff
@@ -2529,6 +2529,63 @@ def test_pivot_table_aggfunc_nunique_with_different_values(self):
 
         tm.assert_frame_equal(result, expected)
 
+    def test_pivot_table_index_and_column_keys_with_nan(self, dropna: bool) -> None:
```
We currently don't add type-hints to tests; can you remove these.

rhshadrach commented on pandas/tests/reshape/test_pivot.py:
```diff
@@ -2529,6 +2529,63 @@ def test_pivot_table_aggfunc_nunique_with_different_values(self):
+        3.0  NaN  NaN  NaN  NaN  4.0
+        NaN  0.0  NaN  NaN  NaN  NaN
+        """
+        data = {"row": [None, *range(4)], "col": [*range(4), None], "val": range(5)}
```
Can you add a reference to the linked issue as a comment on the first line here, `# GH#61113`

rhshadrach commented on pandas/core/frame.py:
```diff
@@ -9347,8 +9347,10 @@ def pivot(
             Do not include columns whose entries are all NaN. If True,
-            rows with a NaN value in any column will be omitted before
-            computing margins.
+            - rows with a NA value in any column will be omitted before computing
+            margins,
+            - index/column keys containing NA values will be dropped (see ``dropna``
+            parameter in :method:``DataFrame.groupby``).
```
I don't believe this renders as intended.

![image](https://github.com/user-attachments/assets/15146979-e8f7-4fe1-94b2-9702e3211e36)

it176131 committed:
Merge branch 'pandas-dev:main' into doc_pivot_table

it176131 committed:
modified:   pandas/core/reshape/pivot.py
	- Updated spelling of :method: to :meth: per comment feedback.

it176131 committed:
modified:   pandas/core/frame.py
	- Updated shared docs replacing :method: with :meth: per comment feedback.

it176131 committed:
modified:   pandas/tests/reshape/test_pivot.py
	- Removed docstring for `test_pivot_table_index_and_column_keys_with_nan` per comment feedback.

it176131 committed:
modified:   pandas/tests/reshape/test_pivot.py
	- Renamed `actual` -> `result` per comment feedback.

it176131 committed:
modified:   pandas/tests/reshape/test_pivot.py
	- Removed type hints for `test_pivot_table_index_and_column_keys_with_nan` per comment feedback.

it176131 committed:
modified:   pandas/tests/reshape/test_pivot.py
	- Removed data manipulations from `expected` to prevent dependencies on other pandas methods.

it176131 committed:
modified:   pandas/tests/reshape/test_pivot.py
	- Added comment with GitHub issue reference number.

it176131 committed:
modified:   pandas/core/frame.py
	- Updated _shared_docs["pivot_table"] to match docstring in pandas/core/reshape/pivot.py.

modified:   pandas/core/reshape/pivot.py
	- Updated `dropna` param docstring in `pivot_table` so the dashes convert to a bulleted list.

it176131 committed:
modified:   pandas/core/frame.py
	- Updated dropna param description in `_shared_docs["pivot_table"]` to use bulleted list similar to `parse_dates` param description in `pandas.io.parsers.readers._doc_read_csv_and_table`.

modified:   pandas/core/reshape/pivot.py
	- Updated docstring for `dropna` param in :func:`pivot_table` to use bulleted list similar to `parse_dates` param description in `pandas.io.parsers.readers._doc_read_csv_and_table`.

it176131 committed:
Merge branch 'main' into doc_pivot_table

rhshadrach commented on pandas/core/frame.py:
```diff
@@ -9347,8 +9347,12 @@ def pivot(
-            rows with a NaN value in any column will be omitted before
-            computing margins.
+
+            * rows with a NA value in any column will be omitted before computing
```
nit: `an NA` value, not `a NA` value. This is because the letter `N` is spelled `en` which starts with a vowel.

rhshadrach commented on pandas/core/frame.py:
```diff
@@ -9347,8 +9347,12 @@ def pivot(
+            * rows with a NA value in any column will be omitted before computing
+              margins,
+            * index/column keys containing NA values will be dropped (see ``dropna``
+              parameter in :meth:``DataFrame.groupby``).
```
Needs to be a single backtick, `` :meth:`DataFrame.groupby` ``

rhshadrach commented on pandas/core/reshape/pivot.py:
```diff
@@ -102,8 +102,11 @@ def pivot_table(
-        rows with a NaN value in any column will be omitted before
-        computing margins.
+
+        * rows with a NA value in any column will be omitted before computing margins,
```
`an NA` again

rhshadrach reviewed (changes requested):
Looking good!

it176131 committed:
modified:   pandas/core/frame.py
	- Updated `_shared_docs["pivot_table"]` value per comments.

modified:   pandas/core/reshape/pivot.py
	- Updated `dropna` param docstring in :func:`pivot_table` per comments.

it176131 committed:
Merge branch 'main' into doc_pivot_table

it176131 commented:
@rhshadrach ready for review

rhshadrach reviewed (approved):
lgtm

rhshadrach commented:
Thanks @it176131 - nice work!
